tools/clang/lib/SPIRV/DeclResultIdMapper.cpp

```diff
@@ -166,72 +168,137 @@ class LocationSet {
     return nextLoc++;
   }
 
+  /// Returns true if the given location number is already used.
+  bool isLocUsed(uint32_t loc) { return usedLocs[loc]; }
+
 private:
   llvm::SmallBitVector usedLocs; ///< All previously used locations
   uint32_t nextLoc;              ///< Next available location
 };
 } // namespace
 
-void DeclResultIdMapper::finalizeStageIOLocations() {
-  { // Check semantic duplication
-    llvm::StringSet<> seenInputSemantics;
-    llvm::StringSet<> seenOutputSemantics;
-    bool success = true;
+bool DeclResultIdMapper::checkSemanticDuplication(bool forInput) {
+  llvm::StringSet<> seenSemantics;
+  bool success = true;
+
+  for (const auto &var : stageVars) {
+    auto s = var.getSemanticStr();
+
+    if (forInput && var.getSigPoint()->IsInput()) {
+      if (seenSemantics.count(s)) {
+        emitError("input semantic '%0' used more than once") << s;
+        success = false;
+      }
+      seenSemantics.insert(s);
+    } else if (!forInput && var.getSigPoint()->IsOutput()) {
+      if (seenSemantics.count(s)) {
+        emitError("output semantic '%0' used more than once") << s;
+        success = false;
+      }
+      seenSemantics.insert(s);
+    }
+  }
+
+  return success;
+}
+
+bool DeclResultIdMapper::finalizeStageIOLocations(bool forInput) {
+  if (!checkSemanticDuplication(forInput))
+    return false;
+
+  // Returns true if all non-builtin stage input (if forInput is true) or output
```

Comment:
actually, it returns true only for the given StageVar, not all stage variables  ;-P